### PR TITLE
Make Parser into top-level, opaque type

### PR DESCRIPTION
**Pull request by AntonyBlakey**
This is the enabler for https://github.com/NomicFoundation/slang/issues/241

### CODE REVIEW

File: crates/codegen/parser/src/chumsky/code_generator.rs [hunk 1]
```diff
@@ -166,48 +157,36 @@ impl CodeGenerator {
...
+                        // Macros -------------------------------------------
 
-                            {}
+                        {}
 
-                            // Define all productions ---------------------------
+                        // Define all productions ---------------------------
 
-                            {}
+                        {}
```
Comment (OmarTawfik) [inline crates/codegen/parser/src/chumsky/code_generator.rs:178]: nit: suggestion on using named arguments in long `format!()` expressions, to make it a bit easier to see what is getting added/removed:

```rust
format!("
     // Declare all versions -----------------------------

     {versions}

     // Declare all productions --------------------------

     {productions}
",
    versions = get_versions(),
    productions = get_productions(),
);
```

File: crates/solidity/inputs/schema/topics/01-file-structure/05-trivia.yml [hunk 2]
```diff
@@ -47,7 +47,7 @@
   kind: Token
   delimitedBy:
     open: /*
-    config: { name: content }
+    config: { name: Content }
```
Comment (OmarTawfik) [inline crates/solidity/inputs/schema/topics/01-file-structure/05-trivia.yml:50]: I will add a TODO for myself to add validation for these names. :+1

File: crates/solidity/outputs/rust/lib/src/generated/kinds.rs [hunk 3]
```diff
@@ -296,10 +297,10 @@ pub enum TokenKind {
     ZeroX,
 }
 
-#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize)]
 pub enum RuleKind {
+    ABICoderPragma,
```
Comment (OmarTawfik) [inline crates/solidity/outputs/rust/lib/src/generated/kinds.rs:302]: I wonder what caused that capitalization :O

File: crates/solidity/outputs/rust/lib/src/generated/kinds.rs [hunk 4]
```diff
@@ -569,3 +570,172 @@ pub enum RuleKind {
     YulSwitchStatement,
     YulVariableDeclaration,
 }
+
+#[derive(
+    Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, EnumString, AsRefStr, Display,
+)]
+pub enum ProductionKind {
```
Comment (OmarTawfik) [inline crates/solidity/outputs/rust/lib/src/generated/kinds.rs:577]: Thinking out loud. I worry the additional enum (with duplicaten fields) might cause confusion on user side, and make it harder to correlate between `ProductionKind` and the existing two kinds above.

If this is purely for the sake of the `get_parser()` API, thoughts about creating two APIs instead?

```rust
Language::new(args.version).get_rule_parser(RuleKind::Foo);
Language::new(args.version).get_token_parser(TokenKind::Foo);
```

File: crates/codegen/parser/src/chumsky/code_generator.rs (hunk 1 above)
Comment (OmarTawfik) [reply crates/codegen/parser/src/chumsky/code_generator.rs:178]: If you agree, a few other places in `codegen_parser` can benefit from this. Not blocking for this PR though.

Comment (AntonyBlakey) [reply crates/codegen/parser/src/chumsky/code_generator.rs:178]: sure, agreed

File: crates/solidity/inputs/schema/topics/01-file-structure/05-trivia.yml (hunk 2 above)
Comment (AntonyBlakey) [reply crates/solidity/inputs/schema/topics/01-file-structure/05-trivia.yml:50]: thanks

File: crates/solidity/outputs/rust/lib/src/generated/kinds.rs (hunk 3 above)
Comment (AntonyBlakey) [reply crates/solidity/outputs/rust/lib/src/generated/kinds.rs:302]: I have no idea. I want to revisit this issue to audit naming.

File: crates/solidity/outputs/rust/lib/src/generated/kinds.rs (hunk 4 above)
Comment (AntonyBlakey) [reply crates/solidity/outputs/rust/lib/src/generated/kinds.rs:577]: No, it's somewhat wider impact than this, but I agree, and want to revisit this after the next few test-blocking issues, to clean up the code in general.

Comment (AntonyBlakey) [reply crates/solidity/outputs/rust/lib/src/generated/kinds.rs:577]: [#226] (https://github.com/NomicFoundation/slang/issues/266)

File: crates/codegen/parser/src/chumsky/code_generator.rs (hunk 1 above)
Comment (AntonyBlakey) [reply crates/codegen/parser/src/chumsky/code_generator.rs:178]: [#226] (https://github.com/NomicFoundation/slang/issues/266)

File: crates/solidity/outputs/rust/lib/src/generated/kinds.rs (hunk 4 above)
Comment (AntonyBlakey) [reply crates/solidity/outputs/rust/lib/src/generated/kinds.rs:577]: [#267](https://github.com/NomicFoundation/slang/issues/267)

**Review by OmarTawfik — APPROVED** [review]

**Pull request merged**